lucet-runtime/lucet-runtime-internals/src/instance.rs

```diff
@@ -1119,3 +1204,150 @@ pub(crate) struct EmptyYieldVal;
+            let terminable = state.terminable.swap(false, Ordering::SeqCst);
+
+            if terminable {
+                // we got it! we can signal the instance.
+
+                // Now check what domain the instance is in. We can signal in guest code, but want
+                // to avoid signalling in host code lest we interrupt some function operating on
+                // guest/host shared memory, and invalidate invariants. For example, interrupting
+                // in the middle of a resize operation on a `Vec` could be extremely dangerous.
+                //
+                // Hold this lock through all signalling logic to prevent the instance from
+                // switching domains (and invalidating safety of whichever mechanism we choose here)
+                let mut execution_domain = state.execution_domain.lock().unwrap();
+
+                let result = match *execution_domain {
+                    Domain::Guest => {
+                        let mut curr_tid = state.thread_id.lock().unwrap();
+                        // we're in guest code, so we can just send a signal.
+                        if let Some(thread_id) = *curr_tid {
+                            unsafe {
+                                pthread_kill(thread_id, SIGALRM);
+                            }
+
+                            // wait for the SIGALRM handler to deschedule the instance
+                            //
+                            // this should never actually loop, which would indicate the instance
+                            // was moved to another thread, or we got spuriously notified.
+                            while curr_tid.is_some() {
+                                curr_tid = state.tid_change_notifier.wait(curr_tid).unwrap();
+                            }
+                            Ok(KillSuccess::Signalled)
+                        } else {
+                            panic!("logic error: instance is terminable but not actually running.");
+                        }
+                    }
+                    Domain::Hostcall => {
+                        // the guest is in a hostcall, so the only thing we can do is indicate it
+                        // should terminate and wait.
+                        *execution_domain = Domain::Terminated;
+                        Ok(KillSuccess::Pending)
+                    }
+                    Domain::Terminated => {
+                        // Something else (another KillSwitch?) has already signalled this instance
+                        // to exit when it has completed its hostcall. Nothing to do here.
+                        Err(KillError::NotTerminable)
+                    }
+                };
+                // we must hold the lock on this bool at least until we set the "timed_out" flag.
+                mem::drop(execution_domain);
+                result
+            } else {
+                Err(KillError::NotTerminable)
```

Comment:
Could these be rearranged to use `ok_or`/`?` so they're next to the condition that failed?

Reply:
Done! This was a great improvement.